wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java

```diff
@@ -27,7 +27,11 @@
 @NamedQueries({
   @NamedQuery(
       name = "WorkspaceActivity.getExpired",
-      query = "SELECT a FROM WorkspaceActivity a WHERE a.expiration < :expiration"),
+      query =
+          "SELECT a FROM WorkspaceActivity a WHERE a.expiration < :expiration OR "
+              + "(a.status = org.eclipse.che.api.core.model.workspace.WorkspaceStatus.RUNNING AND "
+              + ":runTimeout > 0 AND "
+              + ":expiration - a.lastRunning > :runTimeout)"),
```

Comment:
1. it would be interesting to see `EXPLAIN` of this query to figure out missing indexes.
2. :runTimeout > 0 - I think it would be better to split this query on two.  Orginal `SELECT a FROM WorkspaceActivity a WHERE a.expiration < :expiration` and current and execute it depending of the value of `runTimeout`
3. if runTimeout > 0 query may look like `WHERE a.expiration < :expiration OR (a.status = org.eclipse.che.api.core.model.workspace.WorkspaceStatus.RUNNING AND a.lastRunning< :X)  I assume it would be easier for DB to optimize. Anyway we need to see `EXPLAIN`

Reply:
Expain results

old
```
                                                    QUERY PLAN                                                    
------------------------------------------------------------------------------------------------------------------
 Index Scan using che_index_ws_activity_expiration on che_workspace_activity a  (cost=0.41..8.10 rows=1 width=82)
   Index Cond: (expiration < '1594228056124'::bigint)
(2 rows)


real	0m0.077s
user	0m0.040s
sys	0m0.020s

```

new
```
                                                                     QUERY PLAN                                                                      
-----------------------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on che_workspace_activity a  (cost=8.88..33.34 rows=2 width=82)
   Recheck Cond: ((expiration < '1594228056124'::bigint) OR ((status)::text = 'RUNNING'::text))
   Filter: ((expiration < '1594228056124'::bigint) OR (((status)::text = 'RUNNING'::text) AND (('1594228056124'::bigint - last_running) > 3600000)))
   ->  BitmapOr  (cost=8.88..8.88 rows=7 width=0)
         ->  Bitmap Index Scan on che_index_ws_activity_expiration  (cost=0.00..4.42 rows=1 width=0)
               Index Cond: (expiration < '1594228056124'::bigint)
         ->  Bitmap Index Scan on che_index_ws_activity_last_stopped  (cost=0.00..4.46 rows=7 width=0)
               Index Cond: ((status)::text = 'RUNNING'::text)
(8 rows)


real	0m0.080s
user	0m0.040s
sys	0m0.018s

```

Reply:
@tomgeorge actually I forgot one important thingy - ideally, we want to have a log message when a workspace is stopped by the `che.limits.workspace.run.timeout` in order to identify and investigate this case which is likely to be an abuser. so @skabashnyuk proposal about splitting query into two is relevant